plugins/org.eclipse.uml2.uml/src/org/eclipse/uml2/uml/util/UMLUtil.java

```diff
@@ -12432,16 +12971,62 @@ public static void setBaseElement(EObject stereotypeApplication,
 		}
 	}
 
+	private static EObject storeStereotypeApplication(
+			org.eclipse.uml2.uml.Package outermostPackage, Element element,
+			EClass definition, Stereotype stereotype) {
+		String storageID = UMLUtil.isStereotypeApplicationStorageEnabled()
+			? getStereotypeApplicationStorageID(outermostPackage)
+			: null;
+		if (storageID != null) {
+			StereotypeApplicationStorage storage = StereotypeApplicationStorage.Registry.INSTANCE
+				.getStereotypeApplicationStorage(storageID);
+			if (storage != null) {
+				EObject stereotypeApplication = EcoreUtil.create(definition);
+				CacheAdapter.getInstance().adapt(stereotypeApplication);
+
+				storage.addStereotypeApplication(element, stereotype,
+					stereotypeApplication);
+
+				basicSetBaseElement(stereotypeApplication, definition, element);
+				return stereotypeApplication;
+			}
+
+			throw new IllegalStateException(
+				"Stereotype application storage not found: " + storageID); //$NON-NLS-1$
+		}
+
+		return null;
+	}
+
 	protected static EObject applyStereotype(Element element, EClass definition,
 			Stereotype stereotype) {
-		return StereotypeApplicationHelper.getInstance(element)
+		Package outermostPackage = getOutermostPackage(element, true);
```

Comment:
Shouldn't this be guarded by isStereotypeApplicationStorageEnabled()?

Reply:
Good catch! There was a guard, but apparently it didn't survive my latest refactoring to further optimize the apply stereotype case, see https://github.com/eclipse-uml2/uml2/pull/2#issuecomment-1865423292

I'll rework this on Wednesday!

Reply:
Ah, but the path to an element's resource set is often shorter than the path to its outermost package's resource set, particularly when the contents of a model are distributed across multiple resources.